### PR TITLE
[SYCL] Implement `annotated_ptr` headers

### PR DESCRIPTION
**Pull request by wangdi4**
Implement `annotated_ptr` headers based on the specification: [sycl_ext_oneapi_annotated_ptr.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc)

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc [hunk 1]
```diff
@@ -561,10 +561,11 @@ namespace sycl::ext::oneapi::experimental {
 template <typename T, typename PropertyListT = properties<>>
 class annotated_ref {
   public:
+    annotated_ref(T*);
```
Comment (gmlueck) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc:564]: Is this part of the API, or is it just an implementation detail?  If the only way to get an `annotated_ref` is from `annotated_ptr`, this should not be part of the API.  Instead, it should be a private constructor and `annotated_ptr` should be a friend.

**Review by gmlueck — COMMENTED** [review]
Assuming this PR fully implements the specification, the specification document should be moved to the "experimental" directory and the Status section should be updated with the appropriate wording from the [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc#status).

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc [hunk 2]
```diff
@@ -48,11 +48,7 @@ the following extensions:
 
 == Status
 
-This is a proposed extension specification, intended to gather community
-feedback. Interfaces defined in this specification may not be implemented yet
-or may be in a preliminary state. The specification itself may also change in
-incompatible ways before it is finalized. *Shipping software products should not
-rely on APIs defined in this specification.*
+This is an experimental extension specification, intended to provide early access to features and gather community feedback. Interfaces defined in this specification are implemented in DPC++, but they are not finalized and may change incompatibly in future versions of DPC++ without prior notice. *Shipping software products should not rely on APIs defined in this specification.*
```
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:51]: Please respect the 80 column limit.  You should be able to cut-and-paste the text (with line breaks) from [template.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc).

**Comment by gmlueck** [comment]
> Assuming this PR fully implements the specification, the specification document should be moved to the "experimental" directory and the Status section should be updated with the appropriate wording from the [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc#status).

Thanks for doing this.  One minor nit: Please update the copyright date on these three extensions to read "2022 - 2023".

**Review by gmlueck — APPROVED** [review]
Spec changes LGTM

**Review by steffenlarsen — COMMENTED** [review]
Overall I think it looks good! Just a couple of small comments and questions:

1. If this means we consider the extensions supported we should also define the feature test macros (`SYCL_EXT_ONEAPI_ANNOTATED_ARG`, `SYCL_EXT_ONEAPI_ANNOTATED_PTR`, and `SYCL_EXT_ONEAPI_KERNEL_ARG_PROPERTIES`) in sycl/source/feature_test.hpp.in.
2. I see some compilation testing but what are the plans for E2E testing? Are you planning on adding some tests to https://github.com/intel/llvm-test-suite?

**Comment by wangdi4** [comment]
> Overall I think it looks good! Just a couple of small comments and questions:
> 
> 1. If this means we consider the extensions supported we should also define the feature test macros (`SYCL_EXT_ONEAPI_ANNOTATED_ARG`, `SYCL_EXT_ONEAPI_ANNOTATED_PTR`, and `SYCL_EXT_ONEAPI_KERNEL_ARG_PROPERTIES`) in sycl/source/feature_test.hpp.in.
> 2. I see some compilation testing but what are the plans for E2E testing? Are you planning on adding some tests to https://github.com/intel/llvm-test-suite?

Base on the offline discussion, we add macros for the new extensions and will open a new PR for E2E tests.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
HIP Timed Out Tests (1):
    SYCL :: Basic/reqd_work_group_size.cpp
Test was disabled for HIP in https://github.com/intel/llvm-test-suite/pull/1625.

**Pull request merged**